### DISCUSSION TITLE
Is it possible for users to 'cancel' pending transactions?

### DISCUSSION THREAD

**Discussion by fbwoolf** [blockstack/stacks-wallet-web#1710] [General]
If users get stuck with a pending transaction, can we provide a way for them to 'replace/cancel' that transaction?

Ethereum provides this documentation as guidance for their users:
https://info.etherscan.com/how-to-cancel-ethereum-pending-transactions/

However, our wallet prevents users from sending 0 STX to themselves, is it possible for us to change this?  If so, do we want to?

**Comment by kyranjamie** [comment]
cc @blockstack/blockchain-team 

Other solutions might include:
- Sending 1 µSTX to null address
- Sending 1 µSTX to another account of the same key

How do we handle the UX when you go to cancel a transaction and it confirms as you're doing it?

**Comment by markmhendrickson** [comment]
This functionality sounds useful in theory, though have we seen particular instances wherein users wanted to cancel transactions outright (vs. those instances when they want to bump up their fees to get them confirmed faster via replace-by-fee)?

**Reply by fbwoolf** [reply]
This came up bc of this: https://discord.com/channels/621759717756370964/745197302255321108/885926460991340585

Our wallet allowed the user to send more STX than they had available which caused that tx to hold up others. It stayed in a pending state for over 24 hrs. To prevent this from happening in the wallet, this issue was created: https://github.com/blockstack/stacks-wallet-web/issues/1701 but needs design consideration.

**Comment by markmhendrickson** [comment]
Here's a reported need [from Discord](https://discord.com/channels/621759717756370964/837416537954320394/887389769586327612) due to simple user error / regret:

![image](https://user-images.githubusercontent.com/28991/133597451-89719c1f-0d5f-431b-823d-7020de69b1c5.png)

**Reply by andresgalante** [reply]
In this case the user sent MIA to his STX wallet in Binance.

**Comment by markmhendrickson** [comment]
This has been coming up more over various Discord channels and Twitter recently. Do we have an idea yet whether this is possible on a protocol level?

**Reply by fbwoolf** [reply]
This hasn't generated much discussion, but there are definitely reports of more users attempting transactions with more STX then they have ...and their tx gets stuck in pending: https://github.com/blockstack/stacks-wallet/issues/940. Our wallet shouldn't allow this to happen or we should address the concept of 'canceling'.

**Comment by markmhendrickson** [comment]
@kantai any insight to the possibility of cancellation (or pseudo cancellation) here?

**Comment by kantai** [comment]
The network doesn't allow you to send 0 stx or to send stx to yourself. However, to cancel a transaction, you can do either of what @kyranjamie suggested: 

*    Sending 1 µSTX to null address
*    Sending 1 µSTX to another account of the same key

**Reply by markmhendrickson** [reply]
Thanks! I'm not sure how these two solutions work exactly. One of you mind explaining?

**Reply by kyranjamie** [reply]
As no cancel functionality exists, the next best thing you can do is change the pending mempool transaction to something inconsequential, such as [these two things].

**Reply by markmhendrickson** [reply]
I get the concept of reducing fees to near zero so the transaction fails, but I'm confused about the `to null address` or `to another account` parts mechanically.

**Reply by fbwoolf** [reply]
I think just meaning this allows you to send the near zero amount to an address that can't be the current account in our case. It will burn to a non existent address or you can send to another account in your wallet to keep the amount.

**Reply by kantai** [reply]
> I get the concept of reducing fees to near zero so the transaction fails, but I'm confused about the to null address or to another account parts mechanically.

The idea isn't that you want the transaction to fail, you are just trying to replace your currently pending transaction with a "do nothing instead" transaction. So let's say I'm trying to buy some NFT with a transaction that uses `nonce = 5`, I then decide against that, so I want to cancel that transaction. To do that, I send another transaction with `nonce = 5`, with a higher transaction fee, that does something negligible (like sending 1 microstx somewhere). The original transaction is replaced out of the mempool by the new transaction, which eventually gets confirmed and included in a block.

**Reply by aulneau** [reply]
> that does something negligible (like sending 1 microstx somewhere)

It could be that the wallet can send to a different address index of the wallet, so the funds are not lost

**Reply by markmhendrickson** [reply]
I see, so it's cancellation via outrunning the initial transaction with the same nonce 🙏 

This suggests that any attempt to cancel a transaction is possible or probable at best? In that the original one could very well still get picked up instead of the subsequent one.

**Reply by whoabuddy** [reply]
It's definitely a game of timing, but can be really helpful especially during periods of high congestion. Generally one of two things will happen (from my experience):

1. The current tx with `nonce = 5` in the mempool goes through while you are trying to RBF, you now have a pending tx with the same `nonce = 5` but the account is `nonce = 6` so future transactions will still go through and it will drop off
2. The new tx replaces the current tx with `nonce = 5`, generates a new txid, and the old tx disappears from explorer but is still visible in  API with a status of `DROPPED_REPLACE_BY_FEE`

I like the idea of automatically generating the 1 uSTX transfer tx to the next wallet index, but even if it went to a burn address I'd find this to be a helpful feature.

Another thing to watch for is that increasing the fee to do the RBF doesn't exceed the wallet balance, otherwise it would be unmineable again anyway. While the "Increase fee" modal does show the current balance it doesn't take into account what the function would be transferring.

**Reply by Hero-Gamer** [reply]
Does RBF HAVE TO HAVE higher tx than original, in your case, if Nonce = 5  is stuck with a fee of 2STX (which actually has happened in the past with users), I wanna replace this, but it's gonna cost me 2.01STX to RBF Nonce = 5 if I wanna take the RBF option, that's a lot. 
And users can have multiples of these 2stx fee stuck, it adds up. 

Is there a workaround this? Like sending to a NULL address with 0.01 STX fee even if it's nonce =5? Any other creative ways?
How about sending a "minus 2STX" Nonce 5 tx, so it cancels the original out or something... just thinking out aloud of silly ideas..Any other creative ways? (wanna put a laughing face but realized no emoji here on GH.... )

@kantai

**Reply by whoabuddy** [reply]
> Does RBF HAVE TO HAVE higher tx than original, in your case, if Nonce = 5 is stuck with a fee of 2STX (which actually has happened in the past with users), I wanna replace this, but it's gonna cost me 2.01STX to RBF Nonce = 5 if I wanna take the RBF option, that's a lot.

Yes, if the fee is equal to or lower than the original the transaction will fail when broadcasted. You can add as little as 0.000001 STX to the existing fee, and also do a 0.000001 STX transfer, making the impact minimized as much as possible.

> And users can have multiples of these 2stx fee stuck, it adds up.

That's why understanding fees, current mempool state, and not setting them too high is important to begin with. Since fees are so cheap now it's easy to decide to send something with a higher fee hoping to have your TX processed first, but once that fee is set there is no reducing it.

It's also why users should avoid sending multiple transactions to complete an action. This may not be something we can prevent, but some of this will be addressed by the discussions in #2129. It really all comes back to user and developer education.

**Comment by gregorycoppola** [comment]
I think explicit "cancel" would be cool.

It's a simpler UI.

Could just have a "cancel" button.

**Reply by sjwhitely** [reply]
+1 - it's a common support request

**Reply by Hero-Gamer** [reply]
plus 1
- if cancelling costs money, as long as if explicitly says how much it costs if you proceed to take the cancel option.
- I don't see any reason why user have to know how it works technically behind the scene.

**Reply by cargocultscience** [reply]
My personal choice would be to have a "cancel" type of transaction that the network would natively support.  To quell possible abuse I do think it makes sense to require that a user pay a fee in excess of the pending transaction they wish to cancel (ie RBF)

Given that adding a new "cancel" transaction type is probably unlikely, I do think that we can build into the UI a "cancel" button that does the RBF + wallet->wallet microstacks transaction behind the scenes.
